submariner/OVN-Interconnect.md

```diff
@@ -0,0 +1,244 @@
+which means it is a non-IC setup.
+
+The RouteAgent  will have these controllers added to it and the one running in gateway node responds to the CRUD operations of Submariner
+endpoints.
+
+#### SubmarinerGWRoute CR Controller
+
+This controller will be responsible for programming the OVN cluster router, it will react only the active g/w node. When a SubmarinerGWRoute
+CR is created or modified the controller shall create or update a routing policy in OVN cluster router with a priority of 20000, and it should
+redirect any traffic destined to remote CIDR to the ovn-k8s-mp0 interface IP.
+
+```bash
+_uuid               : 0459f009-3603-47ac-8ee7-9d958540ed31
+action              : reroute
+external_ids        : {}
+match               : "ip4.dst==10.132.0.0/16"
+nexthop             : []
+nexthops            : ["10.1.1.2"]
+options             : {"external_ids:{submariner"="true}"}
+priority            : 20000
+```
+
+It also programs a route in the ovn-cluster-router, to route the traffic coming from other zones destined to remote cluster IP range via the
+ovn-k8s-mp0 interface IP.
+
+```bash
+_uuid               : d55185d8-3732-45c1-ae90-4a7f8cd191f7
+bfd                 : []
+external_ids        : {}
+ip_prefix           : "10.132.0.0/16"
+nexthop             : "10.1.1.2"
+options             : {}
+output_port         : []
+policy              : []
+route_table         : ""
+```
+
+#### SubmarinerNonGWRoute Controller
+
+This controller will run as part of every route agent pod and connects to the OVN DB. When a SubmarinerNonGWRoute CR is created,
+the route-agent running on the non-GW node will update the ovn-cluster-router with a logical router policy using a priority of 20000
+to send the traffic to the remote cluster via next hop mentioned, which is the transit switch IP to the g/w node. Before adding
+the route it checks if a route exists, if so it skips adding the route again. This is required to prevent duplicate update since there
+can be more than one node in each zone and hence more than one RouteAgent.
+
+```bash
+_uuid               : 22db3005-64c5-4e32-aeb0-642423c30742
+action              : reroute
+external_ids        : {}
+match               : "ip4.dst==10.132.0.0/14"
+nexthop             : []
+nexthops            : ["169.254.0.1"]
+options             : {"external_ids:{submariner"="true}"}
+priority            : 20000
+```
+
+### Backward Compatibility
+
+With this architecture, we are removing the logical switches and routers that Submariner created in the existing implementation.
+During migration, we will need to delete these components and the routes that network-plugin-syncer installed. After that,
+the new controllers shall install the updated routes. The operator needs to be updated as well, not to create a
+network-plugin-syncer pods and remove any existing deployments.
+
+### Gateway FailOver
```

Comment:
Few points regarding Gateway FailOver : 
- this is the naive/worst-case solution, I assume that if for example passive and active gateway nodes are assigned to the same OVN zone, no DP reconfiguration is needed  , right ?
-  Should we recommend to label two nodes that shared OVN zone as gateway nodes?  do we need to update cloud-prepare ? 
- The DP configuration is distributed, do we care about the configuration order?
- Maybe it worth to add that we can consider (as future optimization) using VIP to represent the gateway node switch IP instead of reconfiguring all the routes at non-gw nodes

Reply:
> * is is the naive/worst-case solution, I assume that if for example passive and active gateway nodes are assigned to the same OVN zone, no DP reconfiguration is needed  , right ?
We cannot be sure about the current topology , building intelligence to parse all the nodes and understand it may not be trivial. So despite of same zone or different there will be reconfiguration. Also the ovn-k8s-mp-0 Ip changes when the node changes.
> * Should we recommend to label two nodes that shared OVN zone as gateway nodes?  do we need to update cloud-prepare ?
Not all deployments have multiple nodes in same zone. The common one should be a zone per node.
> * The DP configuration is distributed, do we care about the configuration order?
Order in which it configured does not matter, only when g/w node completes programming other nodes can reach remote cluster.

>Maybe it worth to add that we can consider (as future optimization) using VIP to represent the gateway node switch IP instead of reconfiguring all the routes at non-gw nodes
Sorry missed that will add